File: backend/app/routers/content.py
"""Content endpoints for the wizard.

GET /api/content is readable by any logged-in user (content is global, not
per-user).  All write endpoints require content-admin access (is_admin OR
is_content_admin) and mutate the six content tables introduced in Phase 1.
"""
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from database import get_db
from scaffold.models import (
    User,
    GrantTemplate,
    GrantTypeDef,
    BonusScheduleVariant,
    LoanRate,
    LoanRefinance,
    GrantProgramSettings,
)
from scaffold.auth import get_current_user, get_content_admin_user
from app.content_service import load_content
from schemas import (
    GrantTemplateCreate,
    GrantTemplateUpdate,
    GrantTypeDefCreate,
    GrantTypeDefUpdate,
    BonusScheduleVariantCreate,
    BonusScheduleVariantUpdate,
    LoanRateCreate,
    LoanRateUpdate,
    LoanRefinanceCreate,
    LoanRefinanceUpdate,
    GrantProgramSettingsUpdate,
)

router = APIRouter(prefix="/api/content", tags=["content"])


@router.get("")
def get_content(
    _user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Return the full wizard content blob (grant schedule, loan rates, refi chains, etc.)."""
    return load_content(db)


@router.get("/admin")
def get_admin_content(
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
) -> dict:
    """Return the same content as GET /api/content but as flat lists keyed by row id.

    Used by the content-admin UI so it can edit and delete individual rows.
    """
    templates = db.query(GrantTemplate).order_by(
        GrantTemplate.display_order, GrantTemplate.id
    ).all()
    type_defs = db.query(GrantTypeDef).order_by(
        GrantTypeDef.display_order, GrantTypeDef.name
    ).all()
    variants = db.query(BonusScheduleVariant).order_by(
        BonusScheduleVariant.grant_year,
        BonusScheduleVariant.grant_type,
        BonusScheduleVariant.variant_code,
    ).all()
    rates = db.query(LoanRate).order_by(
        LoanRate.loan_kind, LoanRate.grant_type, LoanRate.year
    ).all()
    refis = db.query(LoanRefinance).order_by(
        LoanRefinance.chain_kind,
        LoanRefinance.grant_year,
        LoanRefinance.grant_type,
        LoanRefinance.order_idx,
    ).all()
    settings_row = db.query(GrantProgramSettings).filter(GrantProgramSettings.id == 1).one_or_none()

    return {
        "grant_templates": [
            {
                "id": t.id,
                "year": t.year,
                "type": t.type,
                "vest_start": t.vest_start,
                "periods": t.periods,
                "exercise_date": t.exercise_date,
                "default_catch_up": bool(t.default_catch_up),
                "show_dp_shares": bool(t.show_dp_shares),
                "display_order": t.display_order,
                "active": bool(t.active),
                "notes": t.notes,
            }
            for t in templates
        ],
        "grant_type_defs": [
            {
                "name": td.name,
                "color_class": td.color_class,
                "description": td.description,
                "is_pre_tax_when_zero_price": bool(td.is_pre_tax_when_zero_price),
                "display_order": td.display_order,
                "active": bool(td.active),
            }
            for td in type_defs
        ],
        "bonus_schedule_variants": [
            {
                "id": v.id,
                "grant_year": v.grant_year,
                "grant_type": v.grant_type,
                "variant_code": v.variant_code,
                "periods": v.periods,
                "label": v.label,
                "is_default": bool(v.is_default),
            }
            for v in variants
        ],
        "loan_rates": [
            {
                "id": r.id,
                "loan_kind": r.loan_kind,
                "grant_type": r.grant_type,
                "year": r.year,
                "rate": r.rate,
                "due_date": r.due_date,
            }
            for r in rates
        ],
        "loan_refinances": [
            {
                "id": r.id,
                "chain_kind": r.chain_kind,
                "grant_year": r.grant_year,
                "grant_type": r.grant_type,
                "orig_loan_year": r.orig_loan_year,
                "order_idx": r.order_idx,
                "date": r.date,
                "rate": r.rate,
                "loan_year": r.loan_year,
                "due_date": r.due_date,
                "orig_due_date": r.orig_due_date,
            }
            for r in refis
        ],
        "grant_program_settings": {
            "loan_term_years": settings_row.loan_term_years if settings_row else 10,
            "latest_rate_year": settings_row.latest_rate_year if settings_row else 2025,
            "dp_shares_start_year": settings_row.dp_shares_start_year if settings_row else 2023,
            "tax_fallback_federal": settings_row.tax_fallback_federal if settings_row else 0.37,
            "tax_fallback_state": settings_row.tax_fallback_state if settings_row else 0.0765,
            "default_purchase_due_month_day_pre2022": settings_row.default_purchase_due_month_day_pre2022 if settings_row else "07-15",
            "default_purchase_due_month_day_post2022": settings_row.default_purchase_due_month_day_post2022 if settings_row else "06-30",
            "price_years_start": settings_row.price_years_start if settings_row else 2018,
            "price_years_end": settings_row.price_years_end if settings_row else 2026,
            "flexible_payoff_enabled": bool(settings_row.flexible_payoff_enabled) if settings_row else False,
        },
    }


def _commit_or_409(db: Session, msg: str = "Constraint violation"):
    try:
        db.commit()
    except IntegrityError:
        db.rollback()
        raise HTTPException(status_code=409, detail=msg)


# ── Grant templates ────────────────────────────────────────────────────────

@router.post("/grant-templates", status_code=201)
def create_grant_template(
    body: GrantTemplateCreate,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = GrantTemplate(**body.model_dump())
    db.add(row)
    _commit_or_409(db, "Grant template already exists for this (year, type)")
    db.refresh(row)
    return {"id": row.id}


@router.put("/grant-templates/{tpl_id}", status_code=200)
def update_grant_template(
    tpl_id: int,
    body: GrantTemplateUpdate,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = db.get(GrantTemplate, tpl_id)
    if not row:
        raise HTTPException(404, "Grant template not found")
    for k, v in body.model_dump(exclude_unset=True).items():
        setattr(row, k, v)
    _commit_or_409(db, "Grant template constraint violation")
    return {"id": row.id}


@router.delete("/grant-templates/{tpl_id}", status_code=204)
def delete_grant_template(
    tpl_id: int,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = db.get(GrantTemplate, tpl_id)
    if not row:
        raise HTTPException(404, "Grant template not found")
    db.delete(row)
    db.commit()


# ── Grant type defs (keyed by name) ────────────────────────────────────────

@router.post("/grant-type-defs", status_code=201)
def create_grant_type_def(
    body: GrantTypeDefCreate,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    if db.get(GrantTypeDef, body.name):
        raise HTTPException(409, "Grant type already exists")
    row = GrantTypeDef(**body.model_dump())
    db.add(row)
    db.commit()
    return {"name": row.name}


@router.put("/grant-type-defs/{name}", status_code=200)
def update_grant_type_def(
    name: str,
    body: GrantTypeDefUpdate,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = db.get(GrantTypeDef, name)
    if not row:
        raise HTTPException(404, "Grant type not found")
    for k, v in body.model_dump(exclude_unset=True).items():
        setattr(row, k, v)
    db.commit()
    return {"name": row.name}


@router.delete("/grant-type-defs/{name}", status_code=204)
def delete_grant_type_def(
    name: str,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = db.get(GrantTypeDef, name)
    if not row:
        raise HTTPException(404, "Grant type not found")
    db.delete(row)
    db.commit()


# ── Bonus schedule variants ────────────────────────────────────────────────

@router.post("/bonus-schedule-variants", status_code=201)
def create_bonus_variant(
    body: BonusScheduleVariantCreate,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = BonusScheduleVariant(**body.model_dump())
    db.add(row)
    _commit_or_409(db, "Bonus variant already exists")
    db.refresh(row)
    return {"id": row.id}


@router.put("/bonus-schedule-variants/{vid}", status_code=200)
def update_bonus_variant(
    vid: int,
    body: BonusScheduleVariantUpdate,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = db.get(BonusScheduleVariant, vid)
    if not row:
        raise HTTPException(404, "Bonus variant not found")
    for k, v in body.model_dump(exclude_unset=True).items():
        setattr(row, k, v)
    _commit_or_409(db, "Bonus variant constraint violation")
    return {"id": row.id}


@router.delete("/bonus-schedule-variants/{vid}", status_code=204)
def delete_bonus_variant(
    vid: int,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = db.get(BonusScheduleVariant, vid)
    if not row:
        raise HTTPException(404, "Bonus variant not found")
    db.delete(row)
    db.commit()


# ── Loan rates ─────────────────────────────────────────────────────────────

@router.post("/loan-rates", status_code=201)
def create_loan_rate(
    body: LoanRateCreate,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = LoanRate(**body.model_dump())
    db.add(row)
    _commit_or_409(db, "Loan rate already exists")
    db.refresh(row)
    return {"id": row.id}


@router.put("/loan-rates/{rid}", status_code=200)
def update_loan_rate(
    rid: int,
    body: LoanRateUpdate,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = db.get(LoanRate, rid)
    if not row:
        raise HTTPException(404, "Loan rate not found")
    for k, v in body.model_dump(exclude_unset=True).items():
        setattr(row, k, v)
    # Enforce the same shape rules as create: after the patch, tax rows need grant_type,
    # purchase_original rows need due_date.
    if row.loan_kind == "tax" and not row.grant_type:
        db.rollback()
        raise HTTPException(400, "tax loan rates require a grant_type")
    if row.loan_kind == "purchase_original" and not row.due_date:
        db.rollback()
        raise HTTPException(400, "purchase_original rates require a due_date")
    _commit_or_409(db, "Loan rate constraint violation")
    return {"id": row.id}


@router.delete("/loan-rates/{rid}", status_code=204)
def delete_loan_rate(
    rid: int,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = db.get(LoanRate, rid)
    if not row:
        raise HTTPException(404, "Loan rate not found")
    db.delete(row)
    db.commit()


# ── Loan refinances ────────────────────────────────────────────────────────

@router.post("/loan-refinances", status_code=201)
def create_loan_refinance(
    body: LoanRefinanceCreate,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = LoanRefinance(**body.model_dump())
    db.add(row)
    db.commit()
    db.refresh(row)
    return {"id": row.id}


@router.put("/loan-refinances/{rid}", status_code=200)
def update_loan_refinance(
    rid: int,
    body: LoanRefinanceUpdate,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = db.get(LoanRefinance, rid)
    if not row:
        raise HTTPException(404, "Loan refinance not found")
    for k, v in body.model_dump(exclude_unset=True).items():
        setattr(row, k, v)
    db.commit()
    return {"id": row.id}


@router.delete("/loan-refinances/{rid}", status_code=204)
def delete_loan_refinance(
    rid: int,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = db.get(LoanRefinance, rid)
    if not row:
        raise HTTPException(404, "Loan refinance not found")
    db.delete(row)
    db.commit()


# ── Grant program settings (singleton) ─────────────────────────────────────

@router.put("/grant-program-settings", status_code=200)
def update_grant_program_settings(
    body: GrantProgramSettingsUpdate,
    _admin: User = Depends(get_content_admin_user),
    db: Session = Depends(get_db),
):
    row = db.query(GrantProgramSettings).filter(GrantProgramSettings.id == 1).one_or_none()
    if not row:
        row = GrantProgramSettings(id=1)
        db.add(row)
    for k, v in body.model_dump(exclude_unset=True).items():
        setattr(row, k, v)
    db.commit()
    return {"id": row.id}

File: frontend/src/__tests__/fixtures/content.ts
import type { ContentBlob } from '../../api.ts'

// Matches the seeded Epic values in backend/app/content_service.py.
// Typed as ContentBlob here rather than cast at each use, so the fixture is
// checked against the real API shape instead of asserted to match it.
export const MOCK_CONTENT: ContentBlob = {
  grant_templates: [
    { id: 1, display_order: 0, year: 2018, type: 'Purchase', vest_start: '2020-06-15', periods: 6, exercise_date: '2018-12-31', default_catch_up: true,  show_dp_shares: false, default_purchase_due_date: '2025-07-15', default_tax_due_date: null },
    { id: 2, display_order: 1, year: 2019, type: 'Purchase', vest_start: '2021-06-15', periods: 6, exercise_date: '2019-12-31', default_catch_up: true,  show_dp_shares: false, default_purchase_due_date: '2026-07-15', default_tax_due_date: null },
    { id: 3, display_order: 2, year: 2020, type: 'Purchase', vest_start: '2021-09-30', periods: 5, exercise_date: '2020-12-31', default_catch_up: true,  show_dp_shares: false, default_purchase_due_date: '2025-07-15', default_tax_due_date: null },
    { id: 4, display_order: 3, year: 2020, type: 'Bonus',    vest_start: '2021-09-30', periods: 4, exercise_date: '2020-12-31', default_catch_up: false, show_dp_shares: false, default_purchase_due_date: null,         default_tax_due_date: '2025-07-15' },
    { id: 5, display_order: 4, year: 2020, type: 'Developer Bonus Shares', vest_start: '2022-09-30', periods: 5, exercise_date: '2020-12-31', default_catch_up: false, show_dp_shares: false, default_purchase_due_date: null, default_tax_due_date: null },
    { id: 6, display_order: 5, year: 2021, type: 'Purchase', vest_start: '2022-09-30', periods: 5, exercise_date: '2021-12-31', default_catch_up: true,  show_dp_shares: false, default_purchase_due_date: '2030-07-15', default_tax_due_date: null },
    { id: 7, display_order: 6, year: 2021, type: 'Bonus',    vest_start: '2022-09-30', periods: 3, exercise_date: '2021-12-31', default_catch_up: false, show_dp_shares: false, default_purchase_due_date: null,         default_tax_due_date: null },
    { id: 8, display_order: 7, year: 2021, type: 'Developer Bonus Shares', vest_start: '2022-09-30', periods: 5, exercise_date: '2021-12-31', default_catch_up: false, show_dp_shares: false, default_purchase_due_date: null, default_tax_due_date: null },
    { id: 9, display_order: 8, year: 2022, type: 'Purchase', vest_start: '2023-09-30', periods: 4, exercise_date: '2022-12-31', default_catch_up: false, show_dp_shares: false, default_purchase_due_date: '2031-06-30', default_tax_due_date: null },
    { id: 10, display_order: 9, year: 2022, type: 'Bonus',    vest_start: '2023-09-30', periods: 3, exercise_date: '2022-12-31', default_catch_up: false, show_dp_shares: false, default_purchase_due_date: null,         default_tax_due_date: null },
    { id: 11, display_order: 10, year: 2022, type: 'Free',     vest_start: '2027-09-30', periods: 1, exercise_date: '2022-12-31', default_catch_up: false, show_dp_shares: false, default_purchase_due_date: null,         default_tax_due_date: '2031-06-30' },
    { id: 12, display_order: 11, year: 2023, type: 'Purchase', vest_start: '2024-09-30', periods: 4, exercise_date: '2023-12-31', default_catch_up: false, show_dp_shares: true,  default_purchase_due_date: '2032-06-30', default_tax_due_date: null },
    { id: 13, display_order: 12, year: 2023, type: 'Bonus',    vest_start: '2024-09-30', periods: 3, exercise_date: '2023-12-31', default_catch_up: false, show_dp_shares: false, default_purchase_due_date: null,         default_tax_due_date: null },
    { id: 14, display_order: 13, year: 2024, type: 'Purchase', vest_start: '2025-09-30', periods: 4, exercise_date: '2024-12-31', default_catch_up: false, show_dp_shares: true,  default_purchase_due_date: '2033-06-30', default_tax_due_date: null },
    { id: 15, display_order: 14, year: 2024, type: 'Bonus',    vest_start: '2025-09-30', periods: 3, exercise_date: '2024-12-31', default_catch_up: false, show_dp_shares: false, default_purchase_due_date: null,         default_tax_due_date: null },
    { id: 16, display_order: 15, year: 2025, type: 'Purchase', vest_start: '2026-09-30', periods: 4, exercise_date: '2025-12-31', default_catch_up: false, show_dp_shares: true,  default_purchase_due_date: '2034-06-30', default_tax_due_date: null },
    { id: 17, display_order: 16, year: 2025, type: 'Bonus',    vest_start: '2026-09-30', periods: 3, exercise_date: '2025-12-31', default_catch_up: false, show_dp_shares: false, default_purchase_due_date: null,         default_tax_due_date: null },
  ],
  bonus_schedule_variants: [
    { id: 1, grant_year: 2020, grant_type: 'Bonus', variant_code: 'A', periods: 2, label: 'A (2 years)', is_default: false },
    { id: 2, grant_year: 2020, grant_type: 'Bonus', variant_code: 'B', periods: 3, label: 'B (3 years)', is_default: false },
    { id: 3, grant_year: 2020, grant_type: 'Bonus', variant_code: 'C', periods: 4, label: 'C (4 years)', is_default: true  },
  ],
  loan_rates: {
    interest: { '2020': 0.0086, '2021': 0.0091, '2022': 0.0328, '2023': 0.0437, '2024': 0.037, '2025': 0.0379 },
    tax: {
      'Catch-Up': { '2021': 0.0086, '2022': 0.0187, '2023': 0.0356, '2024': 0.043, '2025': 0.0407 },
      'Bonus':    { '2021': 0.0086, '2022': 0.0293, '2023': 0.0385, '2024': 0.037 },
    },
    purchase_original: {
      '2018': { rate: 0.0307, due_date: '2025-07-15' },
      '2019': { rate: 0.0307, due_date: '2026-07-15' },
      '2020': { rate: 0.0038, due_date: '2025-07-15' },
      '2021': { rate: 0.0086, due_date: '2030-07-15' },
      '2022': { rate: 0.0187, due_date: '2031-06-30' },
      '2023': { rate: 0.0356, due_date: '2032-06-30' },
      '2024': { rate: 0.037,  due_date: '2033-06-30' },
      '2025': { rate: 0.0406, due_date: '2034-06-30' },
    },
  },
  loan_refinances: {
    purchase: {
      '2018': [
        { date: '2020-01-01', rate: 0.0169, loan_year: 2020, due_date: '2025-07-15' },
        { date: '2020-06-01', rate: 0.0043, loan_year: 2020, due_date: '2025-07-15' },
        { date: '2021-11-01', rate: 0.0086, loan_year: 2021, due_date: '2027-07-15' },
      ],
      '2019': [
        { date: '2020-06-01', rate: 0.0043, loan_year: 2020, due_date: '2026-07-15' },
        { date: '2021-11-01', rate: 0.0086, loan_year: 2021, due_date: '2028-07-15' },
      ],
      '2020': [
        { date: '2021-11-01', rate: 0.0086, loan_year: 2021, due_date: '2029-07-15' },
      ],
    },
    tax: {
      '2020-Bonus-2021': [
        { date: '2021-11-01', rate: 0.0086, loan_year: 2021, due_date: '2029-07-15', orig_due_date: '2024-07-15' },
      ],
    },
  },
  grant_program_settings: {
    tax_fallback_federal: 0.37,
    tax_fallback_state: 0.0765,
    dp_min_percent: 0.10,
    dp_min_cap: 20000,
    price_years_start: 2018,
    price_years_end: 2026,
  },
  // The server derives these flat listings from the tables above; nothing in the
  // wizard reads them, but ContentBlob has them, so the fixture must too.
  loan_rates_all: [],
  loan_refinances_all: [],
}
